app/master/build.py

```diff
@@ -148,30 +150,7 @@ def prepare(self, subjob_calculator):
         if not self._preparation_coin.spend():
             raise RuntimeError('prepare() was called more than once on build {}.'.format(self._build_id))
 
-        self._state_machine.trigger(BuildEvent.START_PREPARE)
-        # WIP(joey): Move the following code into a PREPARING state callback
-        #  (so that it won't execute if the build has already been canceled.)
-
-        self._logger.info('Fetching project for build {}.', self._build_id)
-        self.project_type.fetch_project()
-        self._logger.info('Successfully fetched project for build {}.', self._build_id)
-
-        job_config = self.project_type.job_config()
-        if job_config is None:
-            raise RuntimeError('Build failed while trying to parse clusterrunner.yaml.')
-
-        subjobs = subjob_calculator.compute_subjobs_for_build(self._build_id, job_config, self.project_type)
-
-        self._unstarted_subjobs = Queue(maxsize=len(subjobs))  # WIP(joey): Move this into BuildScheduler?
-        self._finished_subjobs = Queue(maxsize=len(subjobs))  # WIP(joey): Remove this and just record finished count.
-
-        for subjob in subjobs:
-            self._all_subjobs_by_id[subjob.subjob_id()] = subjob
-            self._unstarted_subjobs.put(subjob)
-
-        self._timing_file_path = self._project_type.timing_file_path(job_config.name)
-        app.util.fs.create_dir(self._build_results_dir())
-        self._state_machine.trigger(BuildEvent.FINISH_PREPARE)
+        self._state_machine.trigger(BuildEvent.START_PREPARE, subjob_calculator=subjob_calculator)
```

Comment:
See my comment on instantiating the subjob_calculator in the callback. If we do that, don't forge to change the method signature for `def prepare()`